pilot/pkg/model/xds_cache.go

```diff
@@ -230,28 +218,27 @@ func (l *lruCache) Add(entry XdsCacheEntry, token CacheToken, value *discovery.R
 	defer l.mu.Unlock()
 	k := entry.Key()
 	cur, f := l.store.Get(k)
-	toWrite := cacheValue{value: value}
 	if f {
-		if token != cur.(cacheValue).token {
+		// This is the stale or same resource
+		if token <= cur.(cacheValue).token || token <= l.token {
 			// entry may be stale, we need to drop it. This can happen when the cache is invalidated
 			// after we call Get.
 			return
 		}
-		// Otherwise, make sure we write the current token again. We don't change the key on writes; the
-		// same token will be used for a value until its invalidated
-		toWrite.token = cur.(cacheValue).token
-	} else {
-		// This is our first time seeing this; this means it was invalidated recently and this is our
-		// first write, or we forgot to call Get before.
-		return
-	}
-	if l.enableAssertions {
-		if toWrite.token == 0 {
-			panic("token cannot be empty. was Get() called before Add()?")
+		if l.enableAssertions {
+			l.assertUnchanged(k, cur.(cacheValue).value, value)
 		}
-		l.assertUnchanged(k, cur.(cacheValue).value, value)
 	}
+
+	if token == 0 || token < l.token {
+		return
+	}
+
+	// token cannot be empty.
+	toWrite := cacheValue{value: value, token: token}
+
 	l.store.Add(k, toWrite)
+	l.token = token
```

Comment:
I would have thought this is a data race.. how come this doesn't fail tests?

Reply:
l.token is protected, and has this interception 
`	if token == 0 || token < l.token {
		return
	}`

Reply:
forgot the whole thing is locked, thanks